### PR TITLE
[ng] router guide: new lifecycle hooks, OnDeactivate, etc

### PR DESCRIPTION
**Pull request by chalin**
Added to sample code and prose for:
- `OnDeactivate`
- `CanReuse`
- `OnReuse`

Contributes to #361, #367
Corresponding API doc update: https://github.com/dart-lang/angular2/pull/290
Staged at: https://angulardart-org-dev2.firebaseapp.com/angular/guide/router/5

### CODE REVIEW

File: src/angular/guide/router/5.jade [hunk 1]
```diff
@@ -163,6 +163,78 @@ ul
...
+:marked
+  **Refresh the browser** and open the JavaScript console.
+  Now visit the crisis center and select each of the first
+  three crises, one at a time. You'll notice that for each route
+  change, from one crisis to another, a message is printed by
+  the `OnDeactivate` lifecycle method.
+
+  Given the nature of a crisis detail component's responsibilities,
+  it seems wasteful to create a new instance each time. A single instance
+  could be used to handle all crisis detail route instructions.
```
Comment (kwalrath) [inline src/angular/guide/router/5.jade]: delete "be used to"

File: src/angular/guide/router/5.jade [hunk 2]
```diff
@@ -163,6 +163,78 @@ ul
...
+
+  You can let the router know about this by using the _CanReuse_ lifecycle hook.
```
Comment (kwalrath) [inline src/angular/guide/router/5.jade]: "know about this" seems imprecise. How about.

To tell the router that the component instance might be reusable, use the _CanReuse_ lifecycle hook.

(Wait... perhaps `CanReuse` is the interface and `routerCanReuse` is the lifecycle hook?)

File: src/angular/guide/router/5.jade [hunk 3]
```diff
@@ -163,6 +163,78 @@ ul
...
+:marked
+  The hook extracts the crisis `id` from the `next` route instruction parameters.
+  The new `_setCrisis` method is defined from a refactoring of `ngOnInit`:
+
++makeExcerpt('app/crisis_center/crisis_detail_component.dart (ngOnInit)')
+
+:marked
+  **Refresh the browser** and select a few crisis center crises.
+  The view details should now reflect the selected crisis.
+  Also notice how the `OnDeactivate` hook is only called if you
```
Comment (kwalrath) [inline src/angular/guide/router/5.jade]: only called -> called only

(generally, it's more precise when "only" is as late as possible.)

File: src/angular/guide/router/5.jade [hunk 4]
```diff
@@ -163,6 +163,78 @@ ul
...
+:marked
+  ## *OnDeactivate* interface
+
+  Each route is handled by a component instance.
+  Generally, when a change in route is requested, the component instance
+  handling the current route is _deactivated_ and then destroyed
+  before a new component instance is created to handle the new route instruction.
+  
+  If a component implements the `OnDeactivate` interface, then
+  the router will call the component's `routerOnDeactivate` method
```
Comment (chalin) [inline src/angular/guide/router/5.jade]: fixed

File: src/angular/guide/router/5.jade [hunk 5]
```diff
@@ -163,6 +163,78 @@ ul
...
+  before the instance is destroyed.
+
+  This lifecycle method gives component instances an opportunity
+  to peform tasks like cleanup and resource deallocation
```
Comment (chalin) [inline src/angular/guide/router/5.jade]: fixed

File: src/angular/guide/router/5.jade (hunk 1 above)
Comment (chalin) [inline src/angular/guide/router/5.jade]: done

File: src/angular/guide/router/5.jade (hunk 2 above)
Comment (chalin) [inline src/angular/guide/router/5.jade]: Fixed.

And yes, for the sake of brevity, I have been deliberately applying the term "hook" loosely. I think that in this case it is ok. We can discuss further if you'd like.

File: src/angular/guide/router/5.jade (hunk 3 above)
Comment (chalin) [inline src/angular/guide/router/5.jade]: fixed (I did hesitate for that one when I wrote it originally)

**Pull request merged**